### ISSUE TITLE
clush integration

### ISSUE THREAD

**Issue by elfurbe** [hpcng/warewulf#825]
### Version of Warewulf
```# wwctl version
wwctl version:   4.4.0-1.git_afcdb21
rpc version: apiPrefix:"rc1" apiVersion:"1" warewulfVersion:"4.4.0-1.git_afcdb21"
```
### Expected behavior

Call this desired behavior, as this is an enhancement request.

Looking at the external node group sources section of the [clush documentation](https://clustershell.readthedocs.io/en/latest/config.html#external-group-sources), I think it would be desirable to have warewulf provide a clush configuration to make the warewulf node inventory a direct source for clush rather than maintaining another alternative listing, a-la genders or otherwise.

There are descriptions of [slurm group bindings](https://clustershell.readthedocs.io/en/latest/config.html#slurm-group-bindings) in the documentation that demonstrate some of the power of this system for creating dynamic groups to use at clush runtime.

Off-hand, I think dynamic groupings for a number of native warewulf node/profile parameters would be useful, including
 - cluster
 - container
 - profile
 
But some very real self-actualizing power could come from a method for mapping arbitrary user tags as groupings, such that users could provide tags for, and thus run bulk commands based on, things like
 - geographic location
 - in-room location
 - connected network/infiniband switches
 - connected PDUs
 
These are just examples, I presume the concept is made clear by them. My thought is that this use case would be addressed well by adding semi-arbitrary filtering capabilities to `wwctl node list`, but perhaps it would be more desirable to leverage some API endpoint, or even parse the nodes.conf directly. I do not currently presume to know how one would go about achieving this functionality, but it feels like it will most likely require Go coding and I'm not qualified to undertake that work.

### What OS/distro are you running
```# cat /etc/os-release
NAME="Rocky Linux"
VERSION="8.7 (Green Obsidian)"
ID="rocky"
ID_LIKE="rhel centos fedora"
VERSION_ID="8.7"
PLATFORM_ID="platform:el8"
PRETTY_NAME="Rocky Linux 8.7 (Green Obsidian)"
ANSI_COLOR="0;32"
LOGO="fedora-logo-icon"
CPE_NAME="cpe:/o:rocky:rocky:8:GA"
HOME_URL="https://rockylinux.org/"
BUG_REPORT_URL="https://bugs.rockylinux.org/"
ROCKY_SUPPORT_PRODUCT="Rocky-Linux-8"
ROCKY_SUPPORT_PRODUCT_VERSION="8.7"
REDHAT_SUPPORT_PRODUCT="Rocky Linux"
REDHAT_SUPPORT_PRODUCT_VERSION="8.7"
```
### How did you install Warewulf

CIQ RPM repo
```[root@sms init.d]# rpm -qa | grep warewulf
warewulf-4.4.0-1.git_afcdb21.el8.x86_64
[root@sms init.d]# yum info warewulf-4.4.0-1.git_afcdb21.el8.x86_64
Last metadata expiration check: 2:28:40 ago on Fri 19 May 2023 02:44:24 PM MST.
Installed Packages
Name         : warewulf
Version      : 4.4.0
Release      : 1.git_afcdb21.el8
Architecture : x86_64
Size         : 66 M
Source       : warewulf-4.4.0-1.git_afcdb21.el8.src.rpm
Repository   : @System
From repo    : ciq
Summary      : A provisioning system for large clusters of bare metal and/or virtual systems
URL          : https://github.com/hpcng/warewulf
License      : BSD-3-Clause
Description  : Warewulf is a stateless and diskless container operating system provisioning
             : system for large clusters of bare metal and/or virtual systems.

[root@sms init.d]# wwctl version
wwctl version:   4.4.0-1.git_afcdb21
rpc version: apiPrefix:"rc1" apiVersion:"1" warewulfVersion:"4.4.0-1.git_afcdb21"
```

**Comment by griznog** [comment]
I've taken a few runs at parsing `wwctl` to do this and it's a combination of terrible output to parse plus being a moving target as that output format evolves, so I settled on the following approach to map node profiles to clustershell groups which should work on any system where a python 3.6 or later is available. 

`wwclush-helper` (Put in PATH somewhere or use full path in the clustershell cofig file that follows.)
```
#!/usr/bin/env python3

# Author: griznog
# Purpose: Helper to parse WW4 node config for clustershell groups.

import os, sys, argparse
import yaml

def nodes_config(cfgfile):
    with open(os.path.join(cfgfile)) as w:
        data = yaml.load(w, Loader=yaml.Loader)
    return data

if __name__ == "__main__":

    # Parse Command line arguments or set defaults.
    parser = argparse.ArgumentParser(description='Warewulf node config clush helper.')
    group_flags = parser.add_mutually_exclusive_group()
    group_flags.add_argument('--map', '-m', dest='map', type = str, default=None, help = "Return list of node in group MAP")
    group_flags.add_argument('--all', '-a', dest='all', action='store_true', default=False, help = "Return all nodes.")
    group_flags.add_argument('--list', '-l', dest='list', action='store_true', default=False, help = "Return all node GROUPs (profiles).")
    group_flags.add_argument('--reverse', '-r', dest='reverse', type = str, default=False, help = "Return groups for node REVERSE.")

    parser.add_argument('--config', '-c', dest='configfile', default='/etc/warewulf/nodes.conf', help = "Path to nodes.conf file.")

    args = parser.parse_args()

    config = nodes_config(args.configfile)
    nodes = config['nodes']
    profiles = config['nodeprofiles']
    
    if args.map:
        result = [ node for node in nodes.keys() if args.map in nodes[node]['profiles'] ]
    elif args.all:
        result = nodes.keys()
    elif args.list:
        result = profiles.keys()
    elif args.reverse:
        try:
            result = nodes[args.reverse]['profiles']
        except KeyError:
            result = []
    else:
        parser.print_help()
        sys.exit(1)
    
    if result is not None and len(result) > 0:
        print(' '.join(result))
    
```

And for clustershell config, `/etc/clustershell/groups.conf.d/warewulf.conf``:

```
# Additional ClusterShell group source config file
#
# Please see `man 5 groups.conf` for further details.
#

# Warewulf Profile Bindings
[ww, warewulf]
map:  wwclush-helper --map $GROUP || true
all: wwclush-helper --all || true
list: wwclush-helper --list || true
reverse: wwclush-helper --reverse $NODE || true
cache_time: 300
```

Examples in action:

```
[I am root!@frankie:(bare):svn://snowstorm,wwovl(admin):bin]# nodeset -L | egrep '(ww|warewulf)'
@ww:overlay-autofs
@ww:role-compute
@ww:role-failsafe
@ww:role-libvirt
@ww:role-login
@ww:role-management
@ww:role-snowflake
@ww:role-storage
@ww:role-webserver
@ww:role-workstation
@ warewulf:overlay-autofs
@ warewulf:role-compute
@ warewulf:role-failsafe
@ warewulf:role-libvirt
@ warewulf:role-login
@ warewulf:role-management
@ warewulf:role-snowflake
@ warewulf:role-storage
@ warewulf:role-webserver
@ warewulf:role-workstation
[I am root!@frankie:(bare):svn://snowstorm,wwovl(admin):bin]# nodeset -f @ww:role-compute
cpu-a-[1-2],cpu-b-[1-6],cpu-c-[1-4],cpu-vm-[1-2],gpu-a-[1-4],gpu-b-[1-6],gpu-c-1,gpu-d-[1-2],gpu-p9-3,gpu-sm[01-02]-[01-20],svc-vm-globus,svc-vm-ondemand,svc-vm-smrtlink,svc-vm-webgateway
[I am root!@frankie:(bare):svn://snowstorm,wwovl(admin):bin]# nodeset -f @ww:role-storage
storage-a-[01-02],storage-b-[01-02],storage-odb[2,5]-[1-2]
```

Profiles were the easiest thing to use for groups and covered most of my needs, but it might be interesting to add support for parsing other values as groups.

**Comment by griznog** [comment]
Actually as I think about it, it'd be pretty straightforward to add support for the stuff you mentioned to this python script. If I get some time I'll expand my script and config file above to add them.

**Comment by griznog** [comment]
Deleted my previous comments as this obsoletes them and they'd just be confusing for anyone reading this. 

See https://github.com/griznog/warewulf-clustershell-integration for a working helper and clustershell config file that will handle cluster, profiles and node tags as the GROUP sources for creating clustershell groups.

**Comment by griznog** [comment]
At the risk of committing heresy by implying that there's not really a need to rewrite the entire world into a single Go binary, once clustershell understands how to build the groups:

```
[I am root!@frankie:(bare)::~]# wwctl node list `nodeset -e @wwtag:SEARCH=czbiohub.org`
NODE NAME              PROFILES                   NETWORK
================================================================================
protospacex            role-snowflake             enp1s0:--
rconnect               role-snowflake             enp1s0:--
```

**Comment by elfurbe** [comment]
The fundamental conclusion here seems to me to be that clush support is not in scope, so I will close the issue as not planned to save us from further workaround discussion. I will open a new issue focused on restoring Warewulf3 functionality that allowed for selecting/providing working sets by various properties other than name, as I think that will be a better context to discuss the merits versus tradeoffs of the capability.

**Comment by mslacken** [comment]
Leave it open right now, as I have to lock in how warewulf can integrate with clush.

**Comment by elfurbe** [comment]
I framed this wrong, the replies made me realize I framed it wrong, that's why I closed it. I continue to think it should be closed.

The fundamental ask of clustershell's group config model is for commands that can return a list of nodes, so commands that can return a list of nodes, as described in #835, is the better thing for warewulf4 to consider. If the capabilities described in #835 already existed, this would not be a github issue for the warewulf project, it would be a PR to the clustershell project adding an [example config](https://github.com/cea-hpc/clustershell/tree/master/conf/groups.conf.d) that made this support first-party, much like the xcat and ace example configs that are already there, along with a PR against the documentation to get a Warewulf4 section similar to the one for [xCAT](https://clustershell.readthedocs.io/en/latest/config.html#xcat-group-bindings).

The correct orientation for the request in this issue is not "warewulf4 should support clustershell", it's  "clustershell should support warewulf4", like it supports those other node management solutions. The best way to get there, in my opinion, is via something like #835, which as a bonus may also deliver benefit to the warewulf4 user/admin beyond clustershell integration.
